pilot/pkg/serviceregistry/kube/controller/ambientindex.go

```diff
@@ -407,6 +420,154 @@ func (c *Controller) setupIndex() *AmbientIndex {
 			}
 		},
 	}
+
+	idx.servicesMap = make(map[types.NamespacedName]*model.Service)
+	idx.handleServiceEntry = func(svc *model.Service, event model.Event) {
+		if svc.Attributes.ServiceEntry == nil {
+			// event for e.g. kube svc; ignore
+			return
+		}
+
+		idx.mu.Lock()
+		defer idx.mu.Unlock()
+
+		// We will accrue updates as we update our internal state
+		updates := sets.New[model.ConfigKey]()
+
+		if event != model.EventAdd {
+			c.cleanupOldWorkloadEntriesInlinedOnServiceEntry(svc, updates)
+		}
+
+		serviceEntryNamespacedName := types.NamespacedName{
+			Name:      svc.Attributes.ServiceEntryName,
+			Namespace: svc.Attributes.ServiceEntryNamespace,
+		}
+
+		// Update indexes
+		if event == model.EventDelete {
+			// servicesMap is used when cleaning up old WEs inlined on a SE (i.e., `ServiceEntry.endpoints`)
+			// so we must delete this after we clean up the old WEs. That way we don't miss any auto-allocated
+			// VIPs during cleanup on the idx.byWorkloadEntry[networkAddr] map
+			delete(idx.servicesMap, serviceEntryNamespacedName)
+		} else {
+			// servicesMap is used when constructing workloads so it must be up to date
+			c.ambientIndex.servicesMap[serviceEntryNamespacedName] = svc
+		}
+
+		pods := c.podsClient.List(metav1.NamespaceAll, klabels.Everything())
+		wls := make(map[string]*model.WorkloadInfo, len(pods))
+		for _, pod := range pods {
+			newWl := c.extractWorkload(pod)
+			if newWl != nil {
+				// Update the pod, since it now has new VIP info
+				networkAddrs := networkAddressFromWorkload(newWl)
+				for _, networkAddr := range networkAddrs {
+					c.ambientIndex.byPod[networkAddr] = newWl
+				}
+				c.ambientIndex.byUID[c.generatePodUID(pod)] = newWl
+				updates.Insert(model.ConfigKey{Kind: kind.Address, Name: newWl.ResourceName()})
+				wls[newWl.Uid] = newWl
+			}
+		}
+
+		workloadEntries := c.getAllControllerWorkloadEntries()
```

Comment:
I don't really understand this logic. Shouldn't it be more like `getWorkloadEntriesInService`, possibly making that func reusable with labels.Selector?

Reply:
addressed in https://github.com/istio/istio/pull/45621/commits/eb05c101ee38455157ab7474ba605be49fead02f